### PR TITLE
Time - Vera

### PR DESCRIPTION
**Pull request by veralizeth**
# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | `all_work` →To show all the work per category. `top_ten` →  To  show the Top ten of each category. `spotlight` → To show the work most voted. 
Describe how you approached testing that model method. What edge cases did you come up with? |  I read all my validations and based on that create each test, second I create a test for all the methods I had on my models. 
What are `session` and `flash`? What is the difference between them? | `Session:` is a hash that persists between HTTP requests using the cookies, without using the database. `Flash:` same behaves but only persist for only one request, after redirects. 
What was one thing that you gained more clarity on through this assignment? | Session and flash, also model methods and how to test them; a little bit of partials(Since I did not use them a lot), and Fixtures and how to use them on tests. 
What is the Heroku URL of your deployed application? | https://serene-inlet-75918.herokuapp.com/

### CODE REVIEW

File: test/controllers/works_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        category: "book",
+        title: "Harry Potter and the Prisoner of Azkaban",
+        creator: "J.K.Rowling",
+        publication_year: "1999",
+        description: "Harry is back at the Dursley's for the summer holidays, where he sees on Muggle television that a convict named Sirius Black has escaped from prison.",
+      },
+    }
+
+    it "can create a work" do
+      expect {
```
Comment (beccaelenzil) [inline test/controllers/works_controller_test.rb:65]: This test failed:

```
"Work.count" didn't change by 1.
        Expected: 7
          Actual: 6
        test/controllers/works_controller_test.rb:65:in `block (3 levels) in <main>'
```

Comment (beccaelenzil) [reply test/controllers/works_controller_test.rb:65]: I can't see the error offhand, but I'm happy to troubleshoot this with you.

File: test/controllers/works_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        patch work_path(work.id), params: new_work_hash
+      }.wont_change "Work.count"
+
+      work.reload
+      must_respond_with :bad_request
+      expect(work.title).wont_be_nil
+    end
+  end
+  describe "destroy" do
+    it "can delete a work" do
```
Comment (beccaelenzil) [inline test/controllers/works_controller_test.rb:142]: This test fails:

```
Minitest::UnexpectedError:         ActionController::UrlGenerationError: No route matches {:action=>"show", :controller=>"works", :id=>nil}, missing required keys: [:id]
            test/controllers/works_controller_test.rb:146:in `block (4 levels) in <main>'
            test/controllers/works_controller_test.rb:145:in `block (3 levels) in <main>'
```

File: app/views/works/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        </tr>
+      </tbody>
+      <%end %>
+  </table>
+  <%else%>
+    <h3> No works! </h3>
+  <%end%>
+</section>
+
+<section class="media-table">
```
Comment (beccaelenzil) [inline app/views/works/index.html.erb:33]: Consider how you could DRY up this code using a loop or a partial view.

File: app/controllers/works_controller.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  def index # Fleshing out our index action.
+    @works = Work.all
+  end
+
+  def show
+    # Raise an exception.
+    # Be aware that any string will be converted to 0
+    work_id = params[:id].to_i
+    # It needs to handle this exception.
+    @work = Work.find_by(id: work_id)
```
Comment (beccaelenzil) [inline app/controllers/works_controller.rb:11]: This functionality would be good for a controller filter.

File: test/models/work_test.rb [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      # Fake description of 201 a's
+      new_work.description = "a" * 201
+
+      # Assert
+      expect(new_work.valid?).must_equal false
+      expect(new_work.errors.messages[:description]).must_equal ["200 characters is the maximum allowed"]
+    end
+  end
+
+  describe "Custom model methods" do
```
Comment (beccaelenzil) [inline test/models/work_test.rb:94]: Consider edge cases for each of these methods. What if there are no works for a particular category. What if two works are tied?

**Comment by beccaelenzil** [comment]
# Media Ranker








## Functional Requirements: Manual Testing


| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️


## Major Learning Goals/Code Review


| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️I appreciate your take on the styling :)
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️


### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code


| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | See in-line comment
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | A method that would find the current user and another that would set the flash warning to "You must be logged in to do that" would be good controller filters, as this functionality is required by multiple actions.
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | See in-line comment in works controller.
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️


### Testing Rails Apps


| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | See in-line comment about edge cases.


## Overall Feedback


| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️










## Code Style Bonus Awards





Was the code particularly impressive in code style for any of these reasons (or more...?)


| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by beccaelenzil — COMMENTED** [review]
Good work overall. Your website implements all the key functionality and you've added your unique style while still matching the original site closely. You've also done a great job writing custom model methods. I've left a few comments about using partial views, controller filters, and making sure to test edge cases for custom model methods. Keep up the hard work!
